File: 2 - Work/Phase 1 Parsing/Phase_1.R
# Purpose: Master pipeline - parse, spatial-join, proxy-merge, and classify
#          baseline land values for all US golf courses (Phase 1).
# Inputs:  00 - Data Sources/Original Data/Golf Courses-USA.csv
#          00 - Data Sources/Original Data/2022 - USDA County Data - Ag Use.csv
#          00 - Data Sources/Original Data/2024 - FHFA June 20 Land Prices.xlsx
#          https://www.ers.usda.gov/media/5768/2023-rural-urban-continuum-codes.csv
# Outputs: Phase 1 Parsing/Data/R/R_Phase1_Parsed_Golf_Courses.csv
#          Phase 1 Parsing/Data/R/R_Phase1_Spatial_Joined_Golf_Courses.csv
#          Phase 1 Parsing/Data/R/R_Phase1_Valuation_Joined_Golf_Courses.csv
#          Phase 1 Parsing/Data/R/R_Phase1_Baseline_Golf_Valuation.csv


# === 1. LIBRARIES ===

suppressPackageStartupMessages({
  library(wooldridge)
  library(tidyverse)
  library(readxl)
  library(sf)
  library(tigris)
  library(future)
  library(furrr)
  library(parallelly)
  library(this.path)
})


# === 2. GLOBALS & PATHS ===

options(tigris_use_cache = TRUE)

SCRIPT_DIR   <- this.path::this.dir()
ROOT_DIR     <- file.path(SCRIPT_DIR, "..")
DATA_DIR     <- file.path(ROOT_DIR, "00 - Data Sources", "Original Data")
OUTPUT_DIR   <- file.path(SCRIPT_DIR, "Data", "R")

SAFE_WORKERS <- min(availableCores() - 6, 20)
plan(multisession, workers = SAFE_WORKERS)

if (!dir.exists(OUTPUT_DIR)) dir.create(OUTPUT_DIR, recursive = TRUE)

RAW_CSV      <- file.path(DATA_DIR, "Golf Courses-USA.csv")
USDA_IN      <- file.path(DATA_DIR, "2022 - USDA County Data - Ag Use.csv")
FHFA_IN      <- file.path(DATA_DIR, "2024 - FHFA June 20 Land Prices.xlsx")
RUCC_URL     <- "https://www.ers.usda.gov/media/5768/2023-rural-urban-continuum-codes.csv?v=98246"

OUT_PARSED   <- file.path(OUTPUT_DIR, "R_Phase1_Parsed_Golf_Courses.csv")
OUT_SPATIAL  <- file.path(OUTPUT_DIR, "R_Phase1_Spatial_Joined_Golf_Courses.csv")
OUT_VAL_JOIN <- file.path(OUTPUT_DIR, "R_Phase1_Valuation_Joined_Golf_Courses.csv")
OUT_BASELINE <- file.path(OUTPUT_DIR, "R_Phase1_Baseline_Golf_Valuation.csv")


# === 3. FUNCTIONS ===


# === 4. EXECUTION ===

for (path in c(RAW_CSV, USDA_IN, FHFA_IN)) {
  if (!file.exists(path)) stop(paste("Input file not found:", path))
}

# PART A: Parse & Deduplicate
cat(" 1  Loading and Parsing raw Golf Courses CSV\n")
courses_df <- read_csv(
  RAW_CSV,
  col_names = c("Longitude", "Latitude", "Name_State", "Details"),
  show_col_types = FALSE
)
original_n <- nrow(courses_df)

courses_df <- courses_df |>
  mutate(
    Course_Name    = str_remove(Name_State, "-.*$"),
    State_Abbr     = str_extract(Name_State, "[A-Z]{2}$"),
    Ownership_Type = str_remove_all(str_extract(Details, "^\\([^)]+\\)"), "[()]"),
    Holes          = as.numeric(str_extract(Details, "(?<=\\()\\d+(?= Holes\\))")),
    Zip_Code       = str_extract(Details, paste0("(?<=", State_Abbr, " )\\d{5}")),
    City           = str_remove(str_remove(str_extract(Details, paste0(",([^,]+),", State_Abbr)), "^,\\s*"), paste0(",", State_Abbr)),
    Address        = str_remove(str_remove(str_extract(Details, "\\), (.*?),(?=\\s*[^,]+,[A-Z]{2})"), "^\\), "), ",$")
  ) |>
  select(Longitude, Latitude, Course_Name, State_Abbr, Ownership_Type, Holes, Address, City, Zip_Code)

cat(" 2  Deduplicating and Generating IDs\n")
courses_df <- courses_df |>
  mutate(Lat_Round = round(Latitude, 4), Lon_Round = round(Longitude, 4)) |>
  group_by(Lat_Round, Lon_Round, Course_Name) |>
  arrange(desc(Holes)) |>
  slice(1) |>
  ungroup() |>
  mutate(course_id = row_number()) |>
  select(course_id, Course_Name, Ownership_Type, Holes, Address, City, State_Abbr, Zip_Code, Longitude, Latitude)

write_csv(courses_df, OUT_PARSED)
cat(sprintf("    [OK] Parsed data saved -> %s\n", OUT_PARSED))


# PART B: Spatial Join
cat(" 3  Converting to sf object (EPSG:4326)\n")
courses_sf <- courses_df |>
  filter(!is.na(Longitude), !is.na(Latitude)) |>
  st_as_sf(coords = c("Longitude", "Latitude"), crs = 4326, remove = FALSE)

cat(" 4  Downloading 2022 US County boundaries (tigris)\n")
# [METHODOLOGY] CRS: EPSG 4326 (WGS 84) - projects county boundaries to match golf course point CRS for spatial join
county_sf <- counties(cb = TRUE, year = 2022, resolution = "20m", progress_bar = FALSE) |>
  st_transform(4326)

cat(" 5  Spatial point-in-polygon join\n")
courses_sf <- st_join(courses_sf, county_sf, join = st_intersects)  # [METHODOLOGY]

data("fips_codes", package = "tigris")
state_map_df <- unique(fips_codes[, c("state_code", "state")])
courses_sf   <- left_join(courses_sf, state_map_df, by = c("STATEFP" = "state_code"))

courses_df <- st_drop_geometry(courses_sf) |>
  mutate(
    FIPS              = str_pad(as.character(GEOID), width = 5, pad = "0"),
    County_Name       = NAME,
    Tigris_State_Abbr = state
  ) |>
  select(course_id, Course_Name, Ownership_Type, Holes, Address, City, State_Abbr,
         Zip_Code, Longitude, Latitude, FIPS, County_Name, Tigris_State_Abbr)

write_csv(courses_df, OUT_SPATIAL)
cat(sprintf("    [OK] Spatial data saved -> %s\n", OUT_SPATIAL))


# PART C: Economic Proxy Merge
cat(" 6  Processing USDA County Ag-Use data\n")
usda_df <- read_csv(USDA_IN, show_col_types = FALSE) |>
  filter(`Data Item` == "AG LAND, INCL BUILDINGS - ASSET VALUE, MEASURED IN $ / ACRE") |>
  mutate(
    FIPS = paste0(
      str_pad(as.integer(`State ANSI`), 2, pad = "0"),
      str_pad(as.integer(`County ANSI`), 3, pad = "0")
    ),
    USDA_Ag_Value_Per_Acre = as.numeric(gsub(",", "", Value))
  ) |>
  filter(!is.na(USDA_Ag_Value_Per_Acre)) |>
  distinct(FIPS, .keep_all = TRUE) |>
  select(FIPS, USDA_Ag_Value_Per_Acre)

cat(" 7  Processing FHFA Panel Counties data\n")
fhfa_df <- read_excel(FHFA_IN, sheet = "Panel Counties", skip = 1) |>
  filter(Year == 2022) |>
  mutate(FIPS = str_pad(as.character(FIPS), width = 5, pad = "0"))

as_is_col <- grep("Per Acre, As-Is", names(fhfa_df), value = TRUE)
fhfa_df <- fhfa_df |>
  mutate(FHFA_Res_Value_Per_Acre = as.numeric(.data[[as_is_col]])) |>
  distinct(FIPS, .keep_all = TRUE) |>
  select(FIPS, FHFA_Res_Value_Per_Acre)

cat(" 8  Left-joining proxies onto golf courses\n")
courses_df <- courses_df |>
  left_join(usda_df, by = "FIPS") |>
  left_join(fhfa_df, by = "FIPS")

write_csv(courses_df, OUT_VAL_JOIN)
cat(sprintf("    [OK] Valuation data saved -> %s\n", OUT_VAL_JOIN))


# PART D: RUCC Classification & Baseline Valuation
cat(" 9  Fetching 2023 RUCC data from USDA ERS\n")
rucc_df <- read_csv(
  RUCC_URL,
  locale = locale(encoding = "latin1"),
  show_col_types = FALSE
) |>
  filter(Attribute == "RUCC_2023") |>
  mutate(
    FIPS      = str_pad(as.character(FIPS), width = 5, pad = "0"),
    RUCC_2023 = as.integer(Value)
  ) |>
  distinct(FIPS, .keep_all = TRUE) |>
  select(FIPS, RUCC_2023)

cat(" 10 Merging RUCC and Classifying Urban/Rural\n")
courses_df <- courses_df |>
  left_join(rucc_df, by = "FIPS") |>
  mutate(
    county_type = case_when(
      RUCC_2023 %in% 1:3 ~ "Urban",
      RUCC_2023 %in% 4:9 ~ "Rural",
      TRUE ~ NA_character_
    ),
    Baseline_Value_Per_Acre = case_when(
      county_type == "Urban" ~ FHFA_Res_Value_Per_Acre,
      county_type == "Rural" ~ USDA_Ag_Value_Per_Acre,
      TRUE ~ NA_real_
    )
  )


n            <- nrow(courses_df)
missing_fips <- sum(is.na(courses_df$FIPS))
usda_hit     <- sum(!is.na(courses_df$USDA_Ag_Value_Per_Acre))
fhfa_hit     <- sum(!is.na(courses_df$FHFA_Res_Value_Per_Acre))
urban        <- sum(courses_df$county_type == "Urban", na.rm = TRUE)
rural        <- sum(courses_df$county_type == "Rural", na.rm = TRUE)
unclassified <- sum(is.na(courses_df$county_type))
missing_base <- sum(is.na(courses_df$Baseline_Value_Per_Acre))
bv           <- courses_df$Baseline_Value_Per_Acre[!is.na(courses_df$Baseline_Value_Per_Acre)]

cat("\n=== OUTPUT STATISTICS ===\n")
cat(sprintf("  Original raw rows:        %s\n",   formatC(original_n, big.mark = ",")))
cat(sprintf("  Total golf courses:       %s  (after deduplication)\n", formatC(n, big.mark = ",")))
cat(sprintf("  Missing FIPS (no county): %s\n",   formatC(missing_fips, big.mark = ",")))
cat(sprintf("  USDA match rate:          %s / %s  (%.2f%%)\n",
            formatC(usda_hit, big.mark = ","), formatC(n, big.mark = ","), 100 * usda_hit / n))
cat(sprintf("  FHFA match rate:          %s / %s  (%.2f%%)\n",
            formatC(fhfa_hit, big.mark = ","), formatC(n, big.mark = ","), 100 * fhfa_hit / n))
cat(sprintf("  Urban courses:            %s\n",   formatC(urban, big.mark = ",")))
cat(sprintf("  Rural courses:            %s\n",   formatC(rural, big.mark = ",")))
cat(sprintf("  Unclassified (no RUCC):   %d\n",   unclassified))
cat(sprintf("  Missing Baseline value:   %s  (MICE imputation target)\n", formatC(missing_base, big.mark = ",")))
cat("\n  Baseline_Value_Per_Acre summary:\n")
cat(sprintf("    Min:    $%14s\n", formatC(min(bv),    format = "f", digits = 2, big.mark = ",")))
cat(sprintf("    Median: $%14s\n", formatC(median(bv), format = "f", digits = 2, big.mark = ",")))
cat(sprintf("    Mean:   $%14s\n", formatC(mean(bv),   format = "f", digits = 2, big.mark = ",")))
cat(sprintf("    Max:    $%14s\n", formatC(max(bv),    format = "f", digits = 2, big.mark = ",")))

write_csv(courses_df, OUT_BASELINE)
cat(sprintf("\n  [OK] Final Baseline saved -> %s\n", OUT_BASELINE))

File: 2 - Work/Phase 1 Parsing/Phase_1.py
# Purpose: Master pipeline - parse, spatial-join, proxy-merge, and classify
#          baseline land values for all US golf courses (Phase 1).
# Inputs:  00 - Data Sources/Original Data/Golf Courses-USA.csv
#          00 - Data Sources/Original Data/2022 - USDA County Data - Ag Use.csv
#          00 - Data Sources/Original Data/2024 - FHFA June 20 Land Prices.xlsx
#          https://www.ers.usda.gov/media/5768/2023-rural-urban-continuum-codes.csv
# Outputs: Phase 1 Parsing/Data/Python/Py_Phase1_Spatial_Joined_Golf_Courses.csv
#          Phase 1 Parsing/Data/Python/Py_Phase1_Valuation_Joined_Golf_Courses.csv
#          Phase 1 Parsing/Data/Python/Py_Phase1_Baseline_Golf_Valuation.csv


# === 1. LIBRARIES ===

import re
from pathlib import Path

import numpy as np
import pandas as pd
import geopandas as gpd
from pygris import counties


# === 2. GLOBALS & PATHS ===

SCRIPT_DIR = Path(__file__).parent
ROOT_DIR   = SCRIPT_DIR.parent
DATA_DIR   = ROOT_DIR / "00 - Data Sources" / "Original Data"
OUTPUT_DIR = SCRIPT_DIR / "Data" / "Python"

OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

RAW_CSV  = DATA_DIR / "Golf Courses-USA.csv"
USDA_IN  = DATA_DIR / "2022 - USDA County Data - Ag Use.csv"
FHFA_IN  = DATA_DIR / "2024 - FHFA June 20 Land Prices.xlsx"
RUCC_URL = "https://www.ers.usda.gov/media/5768/2023-rural-urban-continuum-codes.csv?v=98246"

OUT_SPATIAL  = OUTPUT_DIR / "Py_Phase1_Spatial_Joined_Golf_Courses.csv"
OUT_VAL_JOIN = OUTPUT_DIR / "Py_Phase1_Valuation_Joined_Golf_Courses.csv"
OUT_BASELINE = OUTPUT_DIR / "Py_Phase1_Baseline_Golf_Valuation.csv"


STATE_FIPS_TO_ABBR = {
    "01": "AL", "02": "AK", "04": "AZ", "05": "AR", "06": "CA",
    "08": "CO", "09": "CT", "10": "DE", "11": "DC", "12": "FL",
    "13": "GA", "15": "HI", "16": "ID", "17": "IL", "18": "IN",
    "19": "IA", "20": "KS", "21": "KY", "22": "LA", "23": "ME",
    "24": "MD", "25": "MA", "26": "MI", "27": "MN", "28": "MS",
    "29": "MO", "30": "MT", "31": "NE", "32": "NV", "33": "NH",
    "34": "NJ", "35": "NM", "36": "NY", "37": "NC", "38": "ND",
    "39": "OH", "40": "OK", "41": "OR", "42": "PA", "44": "RI",
    "45": "SC", "46": "SD", "47": "TN", "48": "TX", "49": "UT",
    "50": "VT", "51": "VA", "53": "WA", "54": "WV", "55": "WI",
    "56": "WY", "60": "AS", "66": "GU", "69": "MP", "72": "PR",
    "78": "VI",
}


# === 3. FUNCTIONS ===

def extract_ownership(detail_str: str) -> str:
    s = str(detail_str).lower()
    for label in ("public", "private", "municipal", "military", "resort"):
        if label in s:
            return label.title()
    return "Unknown"

def extract_holes(detail_str: str) -> int:
    m = re.search(r"\((\d+)\s*Holes?\)", str(detail_str), re.IGNORECASE)
    return int(m.group(1)) if m else 18


# === 4. EXECUTION ===

def main():
    for path in (RAW_CSV, USDA_IN, FHFA_IN):
        if not path.exists():
            raise FileNotFoundError(f"Input file not found: {path}")

    print(" 1  Loading raw Golf Courses CSV")
    courses_df = pd.read_csv(RAW_CSV, header=None,
                             names=["Longitude", "Latitude", "Course_Name", "Details"])
    print(f"    Rows loaded: {len(courses_df):,}")

    print(" 2  Extracting Ownership_Type & Holes via regex")
    courses_df["Ownership_Type"] = courses_df["Details"].apply(extract_ownership)
    courses_df["Holes"]          = courses_df["Details"].apply(extract_holes)

    print(" 3  Converting to GeoDataFrame (EPSG:4326)")
    courses_df = courses_df.dropna(subset=["Longitude", "Latitude"])
    courses_geo = gpd.GeoDataFrame(
        courses_df,
        geometry=gpd.points_from_xy(courses_df["Longitude"], courses_df["Latitude"]),
        crs="EPSG:4326"
    )

    print(" 4  Downloading 2022 US County boundaries (pygris)")
    # [METHODOLOGY] CRS: EPSG 4326 (WGS 84) - projects county boundaries to match golf course point CRS for spatial join
    county_geo = counties(cb=False, year=2022).to_crs("EPSG:4326")

    county_geo["STUSPS"] = county_geo["STATEFP"].map(STATE_FIPS_TO_ABBR).fillna("")

    print(" 5  Spatial point-in-polygon join")
    courses_geo = gpd.sjoin(  # [METHODOLOGY]
        courses_geo,
        county_geo[["GEOID", "NAME", "STUSPS", "geometry"]],
        how="left",
        predicate="intersects"
    )
    courses_geo = courses_geo.rename(columns={
        "GEOID": "FIPS", "NAME": "County_Name", "STUSPS": "Tigris_State_Abbr"
    })
    if "index_right" in courses_geo.columns:
        courses_geo.drop(columns=["index_right"], inplace=True)

    courses_df = pd.DataFrame(courses_geo.drop(columns=["geometry"]))
    courses_df.to_csv(OUT_SPATIAL, index=False)
    print(f"    [OK] Spatial data saved -> {OUT_SPATIAL}")

    courses_df["FIPS"] = (courses_df["FIPS"].astype(str)
                          .str.replace(r"\.0$", "", regex=True).str.zfill(5))

    print(" 6  Processing USDA County Ag-Use data")
    usda_df = pd.read_csv(USDA_IN)
    usda_df = usda_df[
        usda_df["Data Item"] == "AG LAND, INCL BUILDINGS - ASSET VALUE, MEASURED IN $ / ACRE"
    ].copy()
    usda_df["State ANSI"]  = (usda_df["State ANSI"].astype(str)
                               .str.replace(r"\.0$", "", regex=True).str.zfill(2))
    usda_df["County ANSI"] = (usda_df["County ANSI"].astype(str)
                               .str.replace(r"\.0$", "", regex=True).str.zfill(3))
    usda_df["FIPS"] = usda_df["State ANSI"] + usda_df["County ANSI"]
    usda_df["USDA_Ag_Value_Per_Acre"] = pd.to_numeric(
        usda_df["Value"].astype(str).str.replace(",", ""), errors="coerce"
    )
    usda_df = (usda_df[["FIPS", "USDA_Ag_Value_Per_Acre"]]
               .dropna(subset=["USDA_Ag_Value_Per_Acre"])
               .drop_duplicates(subset=["FIPS"]))

    print(" 7  Processing FHFA Panel Counties data")
    fhfa_df = pd.read_excel(FHFA_IN, sheet_name="Panel Counties", skiprows=1)
    fhfa_df = fhfa_df[fhfa_df["Year"] == 2022].copy()
    fhfa_df["FIPS"] = (fhfa_df["FIPS"].astype(str)
                        .str.replace(r"\.0$", "", regex=True).str.zfill(5))
    as_is_col = "Land Value\n(Per Acre, As-Is)"
    fhfa_df = (fhfa_df[["FIPS", as_is_col]]
               .rename(columns={as_is_col: "FHFA_Res_Value_Per_Acre"})
               .drop_duplicates(subset=["FIPS"]))
    fhfa_df["FHFA_Res_Value_Per_Acre"] = pd.to_numeric(
        fhfa_df["FHFA_Res_Value_Per_Acre"], errors="coerce"
    )

    print(" 8  Left-joining proxies onto golf courses")
    courses_df = (courses_df
                  .merge(usda_df, on="FIPS", how="left")
                  .merge(fhfa_df, on="FIPS", how="left"))
    courses_df.to_csv(OUT_VAL_JOIN, index=False)
    print(f"    [OK] Valuation data saved -> {OUT_VAL_JOIN}")

    print(" 9  Fetching 2023 RUCC data from USDA ERS")
    rucc_df = pd.read_csv(RUCC_URL, encoding="latin1")
    rucc_df = rucc_df[rucc_df["Attribute"] == "RUCC_2023"].copy()
    rucc_df["RUCC_2023"] = pd.to_numeric(rucc_df["Value"], errors="coerce")
    rucc_df["FIPS"] = (rucc_df["FIPS"].astype(str)
                       .str.replace(r"\.0$", "", regex=True).str.zfill(5))
    rucc_df = rucc_df[["FIPS", "RUCC_2023"]].drop_duplicates(subset=["FIPS"])

    print(" 10 Merging RUCC and Classifying Urban/Rural")
    courses_df = courses_df.merge(rucc_df, on="FIPS", how="left")
    courses_df["county_type"] = np.where(
        courses_df["RUCC_2023"].isin([1, 2, 3]), "Urban",
        np.where(courses_df["RUCC_2023"].between(4, 9), "Rural", None)
    )

    print(" 11 Building Baseline_Value_Per_Acre")
    courses_df["Baseline_Value_Per_Acre"] = np.where(
        courses_df["county_type"] == "Urban",  courses_df["FHFA_Res_Value_Per_Acre"],
        np.where(courses_df["county_type"] == "Rural", courses_df["USDA_Ag_Value_Per_Acre"], np.nan)
    )

    n            = len(courses_df)
    missing_fips = courses_df["FIPS"].replace("nan", np.nan).isna().sum()
    usda_hit     = courses_df["USDA_Ag_Value_Per_Acre"].notna().sum()
    fhfa_hit     = courses_df["FHFA_Res_Value_Per_Acre"].notna().sum()
    urban        = (courses_df["county_type"] == "Urban").sum()
    rural        = (courses_df["county_type"] == "Rural").sum()
    unclassified = courses_df["county_type"].isna().sum()
    missing_base = courses_df["Baseline_Value_Per_Acre"].isna().sum()
    bv           = courses_df["Baseline_Value_Per_Acre"].dropna()

    print("\n=== OUTPUT STATISTICS ===")
    print(f"  Total golf courses:       {n:,}")
    print(f"  Missing FIPS (no county): {missing_fips:,}")
    print(f"  USDA match rate:          {usda_hit:,} / {n:,}  ({usda_hit/n:.2%})")
    print(f"  FHFA match rate:          {fhfa_hit:,} / {n:,}  ({fhfa_hit/n:.2%})")
    print(f"  Urban courses:            {urban:,}")
    print(f"  Rural courses:            {rural:,}")
    print(f"  Unclassified (no RUCC):   {unclassified:,}")
    print(f"  Missing Baseline value:   {missing_base:,}  (MICE imputation target)")
    print(f"\n  Baseline_Value_Per_Acre summary:")
    print(f"    Min:    ${bv.min():>14,.2f}")
    print(f"    Median: ${bv.median():>14,.2f}")
    print(f"    Mean:   ${bv.mean():>14,.2f}")
    print(f"    Max:    ${bv.max():>14,.2f}")

    courses_df.to_csv(OUT_BASELINE, index=False)
    print(f"\n  [OK] Final Baseline saved -> {OUT_BASELINE}")


if __name__ == "__main__":
    main()
